add quantileInterpolatedWeighted function

bharatnc opened the pull request:
### Changelog category (leave one):
- New Feature


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add `quantileInterpolatedWeighted`/`quantilesInterpolatedWeighted` functions.

The current `quantileExactWeighted` function returns the exact quantile. Add another weighted quantile function that instead approximates the quantiles:

- sorts the input values and corresponding weights.
- builds a cumulative distribution based on weights.
- performs linear interpolation b/w the weights and the values to compute the quantiles. 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

bharatnc committed:
implement quantileApproximateWeighted

bharatnc committed:
add a test for comparision

bharatnc committed:
fix tests

bharatnc committed:
add comments

bharatnc committed:
fix tests

alexey-milovidov commented:
The name can be confusing, as it can be confused with approximate quantile sketches.
Better to name it quantileExactInterpolated and quantileExactInterpolatedWeighted.
Also the functions quantileExactInclusiveWeighted, quantileExactExclusiveWeighted, quantileExactLowWeighted, quantileExactHighWeighted should be added for consistency.

bharatnc commented:
> The name can be confusing, as it can be confused with approximate quantile sketches. Better to name it quantileExactInterpolated and quantileExactInterpolatedWeighted. Also the functions quantileExactInclusiveWeighted, quantileExactExclusiveWeighted, quantileExactLowWeighted, quantileExactHighWeighted should be added for consistency.

Sure, thank you for the comment. I'm thinking of replacing  `quantile(s)ApproximateWeighted` with `quantile(s)InterpolatedWeighted` in this PR. `quantile(s)InterpolatedWeighted` sounds a little more apt to me (without the word `exact`).

bharatnc committed:
try fix fuzzer error

alexey-milovidov commented:
Ok.

bharatnc committed:
try add some numeric limits checks

bharatnc committed:
handle nan

bharatnc commented:
- [ ]  rename `quantileApproximateWeighted` to `quantileInterpolatedWeighted` everywhere.

(I'll do it finally since it's only a rename)

bharatnc committed:
Merge remote-tracking branch 'upstream/master' into ncb/weighted-quantile-approx

bharatnc committed:
rename func to quantileInterpolatedWeighted

bharatnc committed:
try simplifying limit checks

bharatnc commented:
@evillique the PR is ready for your review, would appreciate your feedback.

bharatnc committed:
Update src/AggregateFunctions/AggregateFunctionQuantile.cpp

Co-authored-by: Nikolay Degterinsky <43110995+evillique@users.noreply.github.com>

bharatnc commented on src/AggregateFunctions/AggregateFunctionQuantile.cpp:
```diff
@@ -82,7 +87,9 @@ constexpr bool supportBigInt()
         std::is_same_v<Function<Float32, false>, FuncQuantileExactWeighted<Float32, false>> ||
-        std::is_same_v<Function<Float32, false>, FuncQuantilesExactWeighted<Float32, false>>;
+        std::is_same_v<Function<Float32, false>, FuncQuantilesExactWeighted<Float32, false>> ||
+        std::is_same_v<Function<Float32, false>, FuncQuantilesInterpolatedWeighted<Float32, false>> ||
```
thank you, updated!

evillique commented:
I'm sorry that it took so long to get back to this PR, could you please update this branch to the latest master and add documentation similar to this:

https://github.com/ClickHouse/ClickHouse/blob/6e5e2d5923375c723d3eca82ff5c479904ddd58b/src/Functions/concatWithSeparator.cpp#L158-L163

bharatnc commented:
@evillique thank you for taking a look again, sure I will rebase and update again today.

bharatnc committed:
Merge remote-tracking branch 'upstream/master' into ncb/weighted-quantile-approx

bharatnc committed:
Merge branch 'master' into ncb/weighted-quantile-approx

bharatnc committed:
fix build failure

bharatnc committed:
add docs for quantileInterpolatedWeighted

bharatnc commented:
> @evillique thank you for taking a look again, sure I will rebase and update again today.

@evillique rebased and fixed the build failures mostly (failing ones seem unrelated but obscure to me). 

For docs, I added a separate markdown for this function. (If you want me to add docs in the way you have recommended - can I do that in an another PR ? Looks like it needs to be done for other quantile and possibly for other aggregate functions as well.)

bharatnc commented:
Interesting, the[ AST fuzzer for ubsan](https://s3.amazonaws.com/clickhouse-test-reports/38252/aa03712b6d5b6371d64b2a95f32d8f5a7a39fd83/fuzzer_astfuzzerubsan//report.html) passed for the previous commit - no code changes.

bharatnc commented on src/AggregateFunctions/QuantileInterpolatedWeighted.h:
```diff
@@ -0,0 +1,304 @@
+
+    /// This ignores overflows that may arise during add, sub and mul operations and doesn't aim to provide exact results.
+    /// Since `the quantileInterpolatedWeighted` function itself relies mainly on approximation, this is fine.
+    UnderlyingType interpolate(Float64 level, Float64 xl, Float64 xr, UnderlyingType yl, UnderlyingType yr) const
```
@evillique was taking a look at the ubsan failures.
Last time I got around ubsan errors for these add, sub, mult and div operations, by using functions like `commom::mulIgnoreOverflow` and `commonLLaddIgnoreOverflow` ... etc. It used to work correctly but it looks like they fail now.  I'm looking for suggestions about the best way to handle overflows in this interpolation that I'm doing.

bharatnc replied:
@evillique fixed to use `NO_SANITIZE_UNDEFINED` on the interpolate function which also simplifies the expression. I also observed similar patterns in a few places in `DateTime`, `Aggregate` functions where overflows and NaNs are expected as a result of arithmetic operations.

The [ubsan tests for AST fuzzer](https://s3.amazonaws.com/clickhouse-test-reports/38252/cda37e73bac1a3c36d2d5b2972ed66262f1be459/fuzzer_astfuzzerubsan//report.html) look good now.

bharatnc committed:
fix test failure

bharatnc commented:
@evillique it would be nice to get this merged if the changes look good to you ? Let me know if you have any more changes that you want me to make.

evillique commented on src/AggregateFunctions/QuantileInterpolatedWeighted.h:
```diff
@@ -0,0 +1,302 @@
+
+        /// Maintain a vector of pair of values and weights for easier sorting and for building
+        /// a cumulative distribution using the provided weights.
+        using Pair = typename std::pair<UnderlyingType, Float64>;
```
`Pair` definition can be brought to the main `private:` section.

evillique commented on src/AggregateFunctions/QuantileInterpolatedWeighted.h:
```diff
@@ -0,0 +1,302 @@
+
+        /// vector for populating and storing the cumulative sum using the provided weights.
+        /// example: [0,1,2,3,4,5] -> [0,1,3,6,10,15]
+        std::vector<Float64> weights_cum_sum;
+
+        bool first = true;
+        for (size_t idx = 0; idx < size; ++idx)
+        {
+            accumulated += value_weight_pairs[idx].second;
+
+            if (first)
+            {
+                weights_cum_sum.push_back(value_weight_pairs[idx].second);
+                first = false;
+            }
+            else
+            {
+                weights_cum_sum.push_back(accumulated);
+            }
+        }
```
Behavior with or without the `first` variable seems to be the same.
```suggestion
        std::vector<Float64> weights_cum_sum;
        weights_cum_sum.reserve(size);

        for (size_t idx = 0; idx < size; ++idx)
        {
            accumulated += value_weight_pairs[idx].second;
            weights_cum_sum.push_back(value_weight_pairs[idx].second);
        }
```

evillique commented on src/AggregateFunctions/QuantileInterpolatedWeighted.h:
```diff
@@ -0,0 +1,302 @@
+    /// Get the `size` values of `levels` quantiles. Write `size` results starting with `result` address.
+    /// indices - an array of index levels such that the corresponding elements will go in ascending order.
+    template <typename T>
+    void getManyImpl(const Float64 * levels, const size_t * indices, size_t num_levels, Value * result) const
```
Similar comments can be applied to the `getManyImpl` function.

evillique commented on src/AggregateFunctions/QuantileInterpolatedWeighted.h:
```diff
@@ -0,0 +1,302 @@
+
+        size_t level_index = 0;
+
+        while (level_index < num_levels)
```
```suggestion
        for (size_t level_index = 0; level_index < num_levels; ++level_index)
```

evillique commented on src/AggregateFunctions/QuantileInterpolatedWeighted.h:
```diff
@@ -0,0 +1,302 @@
+            auto level = levels[indices[level_index]];
+
+            size_t idx = 0;
+            if (size >= 2)
```
`value_weight_pairs[x].second` seem to be sorted. Can we use the binary search here?

bharatnc replied:
Ok sure, tried replacing the linear search part (iterative while loop) with a binary search impl in a way that doesn't change the current behavior.

evillique reviewed (commented):
We should probably also mention quantiles* function here: `docs/en/sql-reference/aggregate-functions/reference/quantiles.md:12`

evillique commented on src/AggregateFunctions/QuantileInterpolatedWeighted.h:
```diff
@@ -0,0 +1,302 @@
+
+        /// calculates a simple cumulative distribution based on weights
+        /// Note: if sum_weight is 0 then subsequent division is inf, so set the value of the exp to 0.
+        for (size_t idx = 0; idx < size; ++idx)
+            value_weight_pairs[idx].second
+                = sum_weight == 0 ? 0 : (weights_cum_sum[idx] - 0.5 * value_weight_pairs[idx].second) / sum_weight;
```
Can weights be negative? Because if not then `sum_weight == 0` necessarily means all `value_weight_pairs[idx].second == 0`

bharatnc replied:
No, only unsigned integers are allowed.

evillique replied:
In this case, it only makes sense to go through elements of `value_weight_pairs` if `sum_weight != 0`
```suggestion
        if (sum_weight != 0)
        {
            for (size_t idx = 0; idx < size; ++idx)
                value_weight_pairs[idx].second = (weights_cum_sum[idx] - 0.5 * value_weight_pairs[idx].second) / sum_weight;
         }
```

bharatnc replied:
sure fixed.

bharatnc committed:
review fixes first round

bharatnc committed:
review fix

bharatnc committed:
try binary search when selecting idx

bharatnc commented:
> 

sure, added.

bharatnc commented:
> We should probably also mention quantiles* function here: `docs/en/sql-reference/aggregate-functions/reference/quantiles.md:12`

sure added.

qoega commented:
@bharatnc WDYT you think is it valid to add non integer weights as well? Like when user has % as weight with 1 as cumulative sum.

bharatnc commented:
> 

Hi, I think it'll be okay to support non integer weights depending on the use case. I chose to stick with integer weights for this implementation of `quantileInterpolatedWeighted` as the `quantileExactWeighted` also currently only supports integer weights. Perhaps it'll be nice to have both the functions allow non integer weights..
